woocommerce/woocommerce-gutenberg-products-block
# Remove `useSelectShippingRate` hook and adjust how local state works for `PackageRates`

While looking at #5734 I've established that we **do** need local state, but we can make some improvements so it's more clear as to why.

If the radio buttons are controlled inputs, without a local state you click on shipping options and get no UI feedback until the request completes and the shipping methods update. This is poor UX.

if the radio buttons are uncontrolled inputs, you get instant feedback when selecting a shipping option, but if the shipping method changes elsewhere (on the server, or on another tab) it becomes out of sync with the data stores.

Therefore, the best solution is to store the checked option in local state, and update it both on click, and when the data stores update post-update.

You can see in this PR I was able to remove `useSelectShippingRate`, remove it's local state, and move the logic to `useSelectShippingRates`.

I've moved the Radio button local state into `PackageRates`. There it just stores the value of the selected option. This also reacts to updates from the server using `useEffect`.

`RadioControl` has also been changed to TypeScript.

Closes #5734

### Testing

How to test the changes in this Pull Request:

1. Ensure your store has multiple shipping methods setup
2. Add something to your cart that requires shipping
2. Go to the cart block
3. Select a shipping option. Confirm the UI updates instantly, and the option remains checked once totals are updated
4. Open 2 tabs with a cart block in each
5. In tab 1, select a shipping method
6. In tab 2, change an item qty. Confirm that once the update completes, the correct shipping method is selected.
